File: _engine/csharp/v0_1_0/Request.cs
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Aifunc;

namespace Aifunc.Engine.Csharp.V0_1_0;

/// <summary>
/// HTTP transport and request/response handling for the model API.
///
/// Uses <see cref="HttpClient"/> — zero external dependencies.
/// </summary>
public static class Request
{
    private static readonly HttpClient SharedClient = CreateClient();

    private static readonly Regex Fence = new(
        @"^```(?:json)?\s*\n?(.*?)\n?```$",
        RegexOptions.Singleline | RegexOptions.Compiled);

    private static HttpClient CreateClient()
    {
        var client = new HttpClient();
        client.Timeout = Timeout.InfiniteTimeSpan; // per-request timeout via CancellationToken
        return client;
    }

    /// <summary>
    /// Builds the <see cref="Types.ModelRequestParams"/> for the given artifact, rendered prompt,
    /// and runtime config.
    /// </summary>
    /// <exception cref="AIFuncException">Thrown if <c>config.Model</c> is not set.</exception>
    public static Types.ModelRequestParams BuildRequest(
        Types.AIFuncArtifact artifact, string prompt, AIFuncConfig config)
    {
        if (string.IsNullOrWhiteSpace(config.Model))
        {
            throw new AIFuncException(
                "AIFuncConfig.model is required when mock mode is disabled");
        }

        var messages = new List<Dictionary<string, string>>
        {
            new()
            {
                ["role"]    = "user",
                ["content"] = prompt,
            },
        };

        var responseFormat = new Dictionary<string, string>
        {
            ["type"] = "json_object",
        };

        var resolved = ResolveModelParams(artifact, config.Model!);

        var temperature = config.Temperature ?? resolved.Temperature ?? LegacyTemperature(artifact);
        var topP        = config.TopP ?? resolved.TopP;
        var maxTokens   = config.MaxTokens ?? resolved.MaxTokens ?? LegacyMaxTokens(artifact);

        return new Types.ModelRequestParams(
            config.Model!, messages, temperature, topP, maxTokens, responseFormat);
    }

    /// <summary>
    /// Sends the request to the model API asynchronously.
    /// </summary>
    /// <returns>The parsed response map, or throws <see cref="AIFuncException"/>.</returns>
    public static async Task<Dictionary<string, object?>> SendRequestAsync(
        AIFuncConfig config, Types.ModelRequestParams parameters, int timeoutMs)
    {
        if (string.IsNullOrWhiteSpace(config.BaseUrl))
        {
            throw new AIFuncException(
                "AIFuncConfig.baseUrl is required when mock mode is disabled");
        }
        if (string.IsNullOrWhiteSpace(config.ApiKey))
        {
            throw new AIFuncException(
                "AIFuncConfig.apiKey is required when mock mode is disabled");
        }

        var baseUrl = config.BaseUrl!.TrimEnd();
        if (baseUrl.EndsWith("/")) baseUrl = baseUrl.Substring(0, baseUrl.Length - 1);
        var endpoint = baseUrl.EndsWith("/chat/completions", StringComparison.Ordinal)
            ? baseUrl
            : baseUrl + "/chat/completions";

        var bodyJson = Json.Stringify(parameters.ToJsonMap());

        using var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(timeoutMs));
        using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Content = new StringContent(bodyJson, Encoding.UTF8, "application/json");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", config.ApiKey);

        HttpResponseMessage response;
        try
        {
            response = await SharedClient.SendAsync(request, cts.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException ex)
        {
            throw new AIFuncException("Model API request timed out after " + timeoutMs + "ms", ex);
        }
        catch (HttpRequestException ex)
        {
            throw new AIFuncException("Model API request failed: " + ex.Message, ex);
        }

        using (response)
        {
            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
            if ((int)response.StatusCode >= 400)
            {
                var preview = body.Length > 500 ? body.Substring(0, 500) : body;
                throw new AIFuncException(
                    "Model API returned " + (int)response.StatusCode + ": " + preview);
            }
            return Json.ParseObject(body);
        }
    }

    /// <summary>
    /// Parses a model response map into the content string, then extracts the
    /// JSON object from it.
    /// </summary>
    /// <exception cref="AIFuncException">
    /// Thrown if the response has no choices or the content is not valid JSON.
    /// </exception>
    public static Dictionary<string, object?> ParseResponse(Dictionary<string, object?> responseMap)
    {
        if (!responseMap.TryGetValue("choices", out var choicesVal)
            || choicesVal is not System.Collections.IList choices
            || choices.Count == 0)
        {
            throw new AIFuncException("Model response contains no choices");
        }

        var first = choices[0];
        if (AsDict(first) is not { } firstMap)
        {
            throw new AIFuncException("Model response choice is not an object");
        }

        if (!firstMap.TryGetValue("message", out var msgVal) || AsDict(msgVal) is not { } msgMap)
        {
            throw new AIFuncException("Model response message is not an object");
        }

        msgMap.TryGetValue("content", out var contentVal);
        var content = contentVal is string cs ? cs.Trim() : "";

        // Strip optional markdown code fence
        var fence = Fence.Match(content);
        if (fence.Success)
        {
            content = fence.Groups[1].Value.Trim();
        }

        try
        {
            var parsed = Json.Parse(content);
            if (parsed is not Dictionary<string, object?> result)
            {
                throw new AIFuncException("Model output is not a JSON object");
            }
            return result;
        }
        catch (Json.JsonParseException e)
        {
            var preview = content.Length > 200 ? content.Substring(0, 200) : content;
            throw new AIFuncException(
                "Failed to parse model output as JSON: " + preview, e);
        }
    }

    // -------------------------------------------------------------------------
    // Model param resolution (mirrors Go/TS/Java provider logic)
    // -------------------------------------------------------------------------

    private sealed class StandardParams
    {
        public double? Temperature { get; }
        public double? TopP { get; }
        public int? MaxTokens { get; }

        public StandardParams(double? temperature, double? topP, int? maxTokens)
        {
            Temperature = temperature;
            TopP        = topP;
            MaxTokens   = maxTokens;
        }
    }

    private static StandardParams ResolveModelParams(Types.AIFuncArtifact artifact, string model)
    {
        var mp = artifact.ModelParams;
        if (mp is null) return new StandardParams(null, null, null);

        if (!mp.TryGetValue("rules", out var rulesVal) || rulesVal is not System.Collections.IList rules)
        {
            return new StandardParams(null, null, null);
        }

        foreach (var ruleObj in rules)
        {
            if (AsDict(ruleObj) is not { } rule) continue;

            if (!rule.TryGetValue("match", out var matchVal) || AsDict(matchVal) is not { } match)
                continue;

            if (!MatchesRule(match, model)) continue;

            if (!rule.TryGetValue("params", out var paramsVal) || AsDict(paramsVal) is not { } parameters)
                continue;

            parameters.TryGetValue("temperature", out var temp);
            parameters.TryGetValue("topP", out var topP);
            parameters.TryGetValue("maxTokens", out var maxTokens);

            return new StandardParams(ToDouble(temp), ToDouble(topP), ToInt(maxTokens));
        }
        return new StandardParams(null, null, null);
    }

    private static bool MatchesRule(Dictionary<string, object?> match, string model)
    {
        match.TryGetValue("model", out var mModel);
        match.TryGetValue("models", out var mModels);
        match.TryGetValue("pattern", out var mPattern);

        if (mModel is string ms && ms == model)
            return true;

        if (mModels is System.Collections.IList modelsList)
        {
            foreach (var v in modelsList)
            {
                if (model.Equals(v)) return true;
            }
        }

        if (mPattern is string pattern && !string.IsNullOrWhiteSpace(pattern))
        {
            return GlobMatch(pattern, model);
        }

        // Empty match = wildcard
        var noModel   = !(mModel is string mStr) || string.IsNullOrWhiteSpace(mStr);
        var noModels  = !(mModels is System.Collections.IList ml) || ml.Count == 0;
        var noPattern = !(mPattern is string pStr) || string.IsNullOrWhiteSpace(pStr);
        return noModel && noModels && noPattern;
    }

    private static bool GlobMatch(string pattern, string value)
    {
        var regex = "^" + Regex.Escape(pattern).Replace("\\*", ".*") + "$";
        return Regex.IsMatch(value, regex);
    }

    private static double? LegacyTemperature(Types.AIFuncArtifact artifact)
    {
        var m = artifact.Model;
        if (m is null) return null;
        m.TryGetValue("temperature", out var v);
        return ToDouble(v);
    }

    private static int? LegacyMaxTokens(Types.AIFuncArtifact artifact)
    {
        var m = artifact.Model;
        if (m is null) return null;
        m.TryGetValue("maxTokens", out var v);
        return ToInt(v);
    }

    private static double? ToDouble(object? v)
    {
        if (v is null) return null;
        if (v is byte or sbyte or short or ushort or int or uint or long or ulong
            or float or double or decimal)
        {
            return Convert.ToDouble(v);
        }
        return null;
    }

    private static int? ToInt(object? v)
    {
        if (v is null) return null;
        if (v is byte or sbyte or short or ushort or int or uint or long or ulong
            or float or double or decimal)
        {
            return Convert.ToInt32(v);
        }
        return null;
    }

    private static Dictionary<string, object?>? AsDict(object? v)
    {
        if (v is Dictionary<string, object?> d) return d;
        if (v is System.Collections.IDictionary raw)
        {
            var converted = new Dictionary<string, object?>();
            foreach (System.Collections.DictionaryEntry e in raw)
            {
                converted[Convert.ToString(e.Key) ?? ""] = e.Value;
            }
            return converted;
        }
        return null;
    }
}

File: _engine/java/v0_1_0/Request.java
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package aifunc._engine.java.v0_1_0;

import aifunc.AIFuncException;
import aifunc.AIFuncConfig;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * HTTP transport and request/response handling for the model API.
 *
 * <p>Uses {@link java.net.http.HttpClient} (Java 11+) — zero external dependencies.
 */
public final class Request {

    private Request() {}

    private static final HttpClient SHARED_CLIENT = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .build();

    private static final Pattern FENCE = Pattern.compile(
            "(?s)^```(?:json)?\\s*\\n?(.*?)\\n?```$");

    /**
     * Builds the {@link Types.ModelRequestParams} for the given artifact, rendered prompt,
     * and runtime config.
     *
     * @throws AIFuncException if {@code config.model} is not set
     */
    @SuppressWarnings("unchecked")
    public static Types.ModelRequestParams buildRequest(
            Types.AIFuncArtifact artifact, String prompt, AIFuncConfig config) {

        if (config.getModel() == null || config.getModel().isBlank()) {
            throw new AIFuncException(
                    "AIFuncConfig.model is required when mock mode is disabled");
        }

        List<Map<String, String>> messages = new ArrayList<>();
        Map<String, String> userMsg = new LinkedHashMap<>();
        userMsg.put("role",    "user");
        userMsg.put("content", prompt);
        messages.add(userMsg);

        Map<String, String> responseFormat = new LinkedHashMap<>();
        responseFormat.put("type", "json_object");

        StandardParams resolved = resolveModelParams(artifact, config.getModel());

        Double temperature = config.getTemperature();
        if (temperature == null) temperature = resolved.temperature;
        if (temperature == null) temperature = legacyTemperature(artifact);

        Double topP = config.getTopP();
        if (topP == null) topP = resolved.topP;

        Integer maxTokens = config.getMaxTokens();
        if (maxTokens == null) maxTokens = resolved.maxTokens;
        if (maxTokens == null) maxTokens = legacyMaxTokens(artifact);

        return new Types.ModelRequestParams(
                config.getModel(), messages, temperature, topP, maxTokens, responseFormat);
    }

    /**
     * Sends the request to the model API asynchronously.
     *
     * @return a future that resolves to the parsed response map, or fails with {@link AIFuncException}
     */
    public static CompletableFuture<Map<String, Object>> sendRequestAsync(
            AIFuncConfig config, Types.ModelRequestParams params, int timeoutMs) {

        if (config.getBaseUrl() == null || config.getBaseUrl().isBlank()) {
            return failedFuture(new AIFuncException(
                    "AIFuncConfig.baseUrl is required when mock mode is disabled"));
        }
        if (config.getApiKey() == null || config.getApiKey().isBlank()) {
            return failedFuture(new AIFuncException(
                    "AIFuncConfig.apiKey is required when mock mode is disabled"));
        }

        String base = config.getBaseUrl().stripTrailing();
        if (base.endsWith("/")) base = base.substring(0, base.length() - 1);
        String endpoint = base.endsWith("/chat/completions")
                ? base : base + "/chat/completions";

        String bodyJson = Json.stringify(params.toJsonMap());

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .timeout(Duration.ofMillis(timeoutMs))
                .header("Content-Type",  "application/json")
                .header("Authorization", "Bearer " + config.getApiKey())
                .POST(HttpRequest.BodyPublishers.ofString(bodyJson))
                .build();

        return SHARED_CLIENT.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(response -> {
                    if (response.statusCode() >= 400) {
                        String preview = response.body();
                        if (preview.length() > 500) preview = preview.substring(0, 500);
                        throw new AIFuncException(
                                "Model API returned " + response.statusCode() + ": " + preview);
                    }
                    return Json.parseObject(response.body());
                });
    }

    /** Java 11–compatible replacement for {@code CompletableFuture.failedFuture} (Java 12+). */
    private static <T> CompletableFuture<T> failedFuture(Throwable ex) {
        CompletableFuture<T> future = new CompletableFuture<>();
        future.completeExceptionally(ex);
        return future;
    }

    /**
     * Parses a model response map into the content string, then extracts the
     * JSON object from it.
     *
     * @throws AIFuncException if the response has no choices or the content is not valid JSON
     */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> parseResponse(Map<String, Object> responseMap) {
        Object choicesVal = responseMap.get("choices");
        if (!(choicesVal instanceof List) || ((List<?>) choicesVal).isEmpty()) {
            throw new AIFuncException("Model response contains no choices");
        }

        Object first = ((List<?>) choicesVal).get(0);
        if (!(first instanceof Map)) {
            throw new AIFuncException("Model response choice is not an object");
        }

        Object msgVal = ((Map<?, ?>) first).get("message");
        if (!(msgVal instanceof Map)) {
            throw new AIFuncException("Model response message is not an object");
        }

        Object contentVal = ((Map<?, ?>) msgVal).get("content");
        String content = (contentVal instanceof String) ? ((String) contentVal).strip() : "";

        // Strip optional markdown code fence
        Matcher fence = FENCE.matcher(content);
        if (fence.matches()) {
            content = fence.group(1).strip();
        }

        try {
            Object parsed = Json.parse(content);
            if (!(parsed instanceof Map)) {
                throw new AIFuncException("Model output is not a JSON object");
            }
            return (Map<String, Object>) parsed;
        } catch (Json.JsonParseException e) {
            String preview = content.length() > 200 ? content.substring(0, 200) : content;
            throw new AIFuncException(
                    "Failed to parse model output as JSON: " + preview, e);
        }
    }

    // -------------------------------------------------------------------------
    // Model param resolution (mirrors Go/TS provider logic)
    // -------------------------------------------------------------------------

    private static final class StandardParams {
        final Double  temperature;
        final Double  topP;
        final Integer maxTokens;

        StandardParams(Double temperature, Double topP, Integer maxTokens) {
            this.temperature = temperature;
            this.topP        = topP;
            this.maxTokens   = maxTokens;
        }
    }

    @SuppressWarnings("unchecked")
    private static StandardParams resolveModelParams(Types.AIFuncArtifact artifact, String model) {
        Map<String, Object> mp = artifact.getModelParams();
        if (mp == null) return new StandardParams(null, null, null);

        Object rulesVal = mp.get("rules");
        if (!(rulesVal instanceof List)) return new StandardParams(null, null, null);

        for (Object ruleObj : (List<?>) rulesVal) {
            if (!(ruleObj instanceof Map)) continue;
            Map<String, Object> rule = (Map<String, Object>) ruleObj;

            Object matchVal = rule.get("match");
            if (!(matchVal instanceof Map)) continue;
            Map<String, Object> match = (Map<String, Object>) matchVal;

            if (!matchesRule(match, model)) continue;

            Object paramsVal = rule.get("params");
            if (!(paramsVal instanceof Map)) continue;
            Map<String, Object> params = (Map<String, Object>) paramsVal;

            return new StandardParams(
                    toDouble(params.get("temperature")),
                    toDouble(params.get("topP")),
                    toInt(params.get("maxTokens")));
        }
        return new StandardParams(null, null, null);
    }

    private static boolean matchesRule(Map<String, Object> match, String model) {
        Object mModel = match.get("model");
        if (mModel instanceof String && ((String) mModel).equals(model)) return true;

        Object mModels = match.get("models");
        if (mModels instanceof List) {
            for (Object v : (List<?>) mModels) {
                if (model.equals(v)) return true;
            }
        }

        Object mPattern = match.get("pattern");
        if (mPattern instanceof String) {
            return globMatch((String) mPattern, model);
        }

        // Empty match = wildcard
        boolean noModel   = !(mModel   instanceof String) || ((String) mModel).isBlank();
        boolean noModels  = !(mModels  instanceof List)   || ((List<?>) mModels).isEmpty();
        boolean noPattern = !(mPattern instanceof String) || ((String) mPattern).isBlank();
        return noModel && noModels && noPattern;
    }

    private static boolean globMatch(String pattern, String value) {
        String regex = "^" + Pattern.quote(pattern).replace("\\*", ".*") + "$";
        return value.matches(regex);
    }

    @SuppressWarnings("unchecked")
    private static Double legacyTemperature(Types.AIFuncArtifact artifact) {
        Map<String, Object> m = artifact.getModel();
        return (m != null) ? toDouble(m.get("temperature")) : null;
    }

    @SuppressWarnings("unchecked")
    private static Integer legacyMaxTokens(Types.AIFuncArtifact artifact) {
        Map<String, Object> m = artifact.getModel();
        return (m != null) ? toInt(m.get("maxTokens")) : null;
    }

    private static Double toDouble(Object v) {
        if (v instanceof Number) return ((Number) v).doubleValue();
        return null;
    }

    private static Integer toInt(Object v) {
        if (v instanceof Number) return ((Number) v).intValue();
        return null;
    }
}
